source/extensions/filters/http/grpc_field_extraction/message_converter/message_converter.cc

```diff
@@ -0,0 +1,159 @@
+  is_first_message_ = false;
+  message_data->setIsFinalMessage(is_final_message);
+  conversions_to_message_data_++;
+
+  ENVOY_LOG_MISC(info, "len(parsing_buffer_)={}", parsing_buffer_.length());
+  if (parsed_output->owned_bytes != nullptr) {
+    ABSL_DCHECK(!parsed_output->needs_more_data);
+    ENVOY_LOG_MISC(info, "len(parsed owned_bytes)={}", parsed_output->owned_bytes->length());
+  }
+  return message_data;
+}
+
+absl::StatusOr<std::vector<std::unique_ptr<StreamMessage>>>
+MessageConverter::accumulateMessages(Envoy::Buffer::Instance& data, bool end_stream) {
+  std::vector<std::unique_ptr<StreamMessage>> messages;
+  while (true) {
+    auto message = accumulateMessage(data, end_stream);
+    if (!message.ok()) {
+      return message.status();
+    }
+    if (*message == nullptr) {
+      return messages;
+    }
+    messages.push_back(std::move(*message));
+    if (messages.back()->isFinalMessage()) {
+      return messages;
+    }
+  }
+}
+
+absl::StatusOr<Envoy::Buffer::InstancePtr>
+MessageConverter::convertBackToBuffer(std::unique_ptr<StreamMessage> message) {
+  ABSL_DCHECK(message != nullptr);
+  conversions_to_envoy_buffer_++;
+  if (conversions_to_envoy_buffer_ > conversions_to_message_data_) {
+    return absl::FailedPreconditionError(absl::StrCat(
+        "Data corruption! Number of conversions to StreamMessage = ", conversions_to_message_data_,
+        ", but this is the ", conversions_to_envoy_buffer_,
+        " convertBackToBuffer call. Perhaps this StreamMessage"
+        "belongs to a different MessageConverter?"));
+  }
+
+  // Edge case handling: If there is no proto body, nothing to convert.
+  Envoy::Buffer::InstancePtr output_message = std::make_unique<Envoy::Buffer::OwnedImpl>();
+  if (message->size() == -1) {
+    return output_message;
+  }
+
+  // Create gRPC frame header and add to output buffer.
+  const uint64_t out_message_size = message->size();
+  auto delimiter = sizeToDelimiter(out_message_size);
+  if (!delimiter.ok()) {
+    return delimiter.status();
+  }
+  output_message->add(*delimiter);
+
+  std::unique_ptr<ZeroCopyInputStream> input_stream =
+      message->message()->CreateZeroCopyInputStream();
+
+  // Do NOT let StreamMessage and the current parsed data go out of scope and
+  // automatically destruct. Manage destruction lazily via lambda capture.
+  auto message_lifetime = std::shared_ptr<StreamMessage>(message.release());
+  auto fragment_releasor =
+      [message_lifetime](const void*, size_t,
+                         const Envoy::Buffer::BufferFragmentImpl* this_fragment) {
+        delete this_fragment;
+      };
+
+  // Data owned in StreamMessage is appended to Envoy's Buffer via external
+  // data reference (no copy).
+  const void* data = nullptr;
+  int size = 0;
+  Envoy::Buffer::BufferFragmentImpl* last_fragment = nullptr;
+
+  while (input_stream->Next(&data, &size)) {
+    last_fragment = new Envoy::Buffer::BufferFragmentImpl(data, size, fragment_releasor);
```

Comment:
It looks a bit like this is a memory leak. It's probably not but I don't think I understand why :) Can you shed some light? addBufferFragment says:

```
   * Add externally owned data into the buffer. No copying is done. fragment is not owned. When
   * the fragment->data() is no longer needed, fragment->done() is called.
   * @param fragment the externally owned data to add to the buffer.
```
So it *looks* like the allocated BufferFragmentImpl will be leaked, but maybe the fragment_releasor solves this issue? I guess this is the comment on line 118?

Also instead of of using last_fragment outside of this while loop, how bout if we use `bool is_empty = true` (or something) and set `is_empty = false` inside the loop then check for `if (is_empty) {` down at line 140. (This is independent from resolving the question about the ownership of the allocated buffer).

Reply:
> So it looks like the allocated BufferFragmentImpl will be leaked, but maybe the fragment_releasor solves this issue? I guess this is the comment on line 118?

Exactly. We have many tests for the buffer lifetime especially outOfScope.

https://screenshot.googleplex.com/6XtwwvQyUpBUA4F

>  how bout if we use bool is_empty = true (or something) and set is_empty = false inside the loop then check for if (is_empty)

Thanks for the suggestion. IMHO, the current way is NIT.

Reply:
" IMHO, the current way is NIT." I'm not sure what that means?

Reply:
Done. never mind.